graphql_compiler/compiler/compiler_frontend.py

```diff
@@ -426,27 +415,44 @@ def _compile_vertex_ast(schema, current_schema_type, ast,
             # to jump back to. Otherwise, the traversal could rewind to an old marked location
             # and might ignore entire stretches of applied filtering.
             if context['marked_location_stack'][-1].num_traverses > 0:
-                location = location.revisit()
+                location = query_metadata_table.revisit_location(location)
+
                 context['location_types'][location] = strip_non_null_from_type(current_schema_type)
                 basic_blocks.append(_mark_location(location))
                 context['marked_location_stack'].pop()
                 new_stack_entry = _construct_location_stack_entry(location, 0)
                 context['marked_location_stack'].append(new_stack_entry)
 
+        if fold_directive:
+            inner_location = location.navigate_to_fold(field_name)
+        else:
+            inner_location = location.navigate_to_subpath(field_name)
+
+        inner_location_info = LocationInfo(
+            parent_location=location,
+            type=field_schema_type.name,
+            coerced_from_type=None,
+            optional_scopes_depth=(
```

Comment:
I don't think that's necessary -- the stack is implicit since the optional root locations are the parent locations of all locations that increment the `optional_scopes_depth` value, which means that it's easy to go from a given location to its optional root location.